wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,345 @@
+//! A lint rule that disallows declaration names with type suffixes.
+
+use std::collections::HashSet;
+
+use wdl_ast::AstToken;
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Document;
+use wdl_ast::Span;
+use wdl_ast::SupportedVersion;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+use wdl_ast::v1::BoundDecl;
+use wdl_ast::v1::Decl;
+use wdl_ast::v1::PrimitiveTypeKind;
+use wdl_ast::v1::Type;
+use wdl_ast::v1::UnboundDecl;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// A rule that identifies declaration names that include their type names as a
+/// suffix.
+#[derive(Debug, Default)]
+pub struct DisallowedDeclarationNameRule;
+
+/// Create a diagnostic for a declaration identifier that contains its type name
+fn decl_identifier_with_type(span: Span, decl_name: &str, type_name: &str) -> Diagnostic {
+    Diagnostic::warning(format!(
+        "declaration identifier '{decl_name}' ends with type name '{type_name}'",
+    ))
+    .with_rule("DisallowedDeclarationName")
+    .with_highlight(span)
+    .with_fix("rename the identifier to not include the type name")
+}
+
+impl Rule for DisallowedDeclarationNameRule {
+    fn id(&self) -> &'static str {
+        "DisallowedDeclarationName"
+    }
+
+    fn description(&self) -> &'static str {
+        "Disallows declaration names that end with their type name"
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Declaration names should not include their type as a suffix. This makes the code more \
+         verbose and often redundant. For example, use 'counter' instead of 'counterInt' or \
+         'is_active' instead of 'isActiveBoolean'."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Style, Tag::Clarity])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::VersionStatementNode,
+            SyntaxKind::InputSectionNode,
+            SyntaxKind::OutputSectionNode,
+            SyntaxKind::BoundDeclNode,
+            SyntaxKind::UnboundDeclNode,
```

Comment:
context: We try to keep each list of "exceptable nodes" minimal, preferring the most specific node when excepting two different nodes would lead to the same behavior.

That being said, this will need to be expanded to include task and workflow definition nodes so that _all_ private declarations of a task or workflow can be "excepted" at once. That would make this list one of our longer lists of exceptable nodes, but each node would cover a different set of declarations.

Please add those two nodes to this list, and then I think we'll have all our bases covered.

Reply:
task and workflow definitions still need to be added here